Plutarch/Api/V1/AssocMap.hs

```diff
@@ -47,3 +122,339 @@ instance
+
+-- | Construct a 'PMap' from a list of key-value pairs, sorted by ascending key data.
+pfromAscList :: (POrd k, PIsData k, PIsData v) => Term (s :: S) (PBuiltinMap k v :--> PMap k v)
+pfromAscList = plam $ (passertSorted #) . pcon . PMap
+
+-- | Assert the map is properly sorted
+passertSorted :: (POrd k, PIsData k, PIsData v) => Term (s :: S) (PMap k v :--> PMap k v)
+passertSorted = phoistAcyclic $
+  plam $ \map ->
+    precList
+      ( \self x xs ->
+          plet (pfromData $ pfstBuiltin # x) $ \k ->
+            plam $ \badKey ->
+              pif
+                (badKey # k)
+                (ptraceError "unsorted map")
+                (self # xs # plam (#< k))
+      )
+      (const $ plam $ const map)
+      # pto map
+      # plam (const $ pcon PFalse)
+
+data MapUnion k v f = MapUnion
+  { merge :: f (PBuiltinMap k v :--> PBuiltinMap k v :--> PBuiltinMap k v)
+  , mergeInsert :: f (PBuiltinPair (PAsData k) (PAsData v) :--> PBuiltinMap k v :--> PBuiltinMap k v :--> PBuiltinMap k v)
+  }
+
+type instance
+  ScottEncoded (MapUnion k v) a =
+    (PBuiltinMap k v :--> PBuiltinMap k v :--> PBuiltinMap k v)
+      :--> (PBuiltinPair (PAsData k) (PAsData v) :--> PBuiltinMap k v :--> PBuiltinMap k v :--> PBuiltinMap k v)
+      :--> a
+
+$(Rank2.TH.deriveAll ''MapUnion)
+
+instance (POrd k, PIsData k, PIsData v, Semigroup (Term s v)) => Semigroup (Term s (PMap k v)) where
+  a <> b = punionWith # plam (<>) # a # b
+
+instance (POrd k, PIsData k, PIsData v, Semigroup (Term s v)) => Monoid (Term s (PMap k v)) where
+  mempty = pempty
+
+{- | Combine two 'PMap's applying the given function to any two values that
+ share the same key.
+-}
+punionWith ::
+  (POrd k, PIsData k, PIsData v) =>
+  Term (s :: S) ((v :--> v :--> v) :--> PMap k v :--> PMap k v :--> PMap k v)
+punionWith = phoistAcyclic $
+  plam $
+    \combine -> punionWithData #$ plam $
+      \x y -> pdata (combine # pfromData x # pfromData y)
+
+{- | Combine two 'PMap's applying the given function to any two data-encoded
+ values that share the same key.
+-}
+punionWithData ::
+  (POrd k, PIsData k, PIsData v) =>
+  Term
+    (s :: S)
+    ( (PAsData v :--> PAsData v :--> PAsData v)
+        :--> PMap k v
+        :--> PMap k v
+        :--> PMap k v
+    )
+punionWithData = phoistAcyclic $
+  plam $ \combine x y ->
+    pcon $ PMap $ mapUnion # combine # field merge # pto x # pto y
+
+mapUnion ::
+  (POrd k, PIsData k, PIsData v) =>
+  Term (s :: S) ((PAsData v :--> PAsData v :--> PAsData v) :--> ScottEncoding (MapUnion k v) (a :: PType))
+mapUnion = plam $ \combine ->
+  letrec $ \MapUnion {merge, mergeInsert} ->
+    MapUnion
+      { merge = plam $ \xs ys -> pmatch xs $ \case
+          PNil -> ys
+          PCons x xs' -> mergeInsert # x # xs' # ys
+      , mergeInsert = plam $ \x xs ys -> pmatch ys $ \case
+          PNil -> pcons # x # xs
+          PCons y ys' ->
+            plet (pfstBuiltin # x) $ \xk ->
+              plet (pfstBuiltin # y) $ \yk ->
```

Comment:
I think you want to `plet` `x` and `y` too. Can't remember whether they already are, but an extra `plet` is free.

Reply:
Do you mean `plet (psndBuiltin # x) $ \xv ->`?

Reply:
Nope, `x` itself

Reply:
I tried; `plet x $ \x->` made no difference but `plet y $ \y->` did to my surprise. I've pushed the commit, but note that while the script size and CPU cost went down the memory cost went up slightly. It still seems to be an improvement.